File: provers/gnark/go/groth16.go
package main

import (
	"encoding/binary"
	"log"
	"sync"

	"github.com/tiannian/rsnark/provers-gnark/circuit"
	"github.com/tiannian/rsnark/provers-gnark/prover"
	"github.com/tiannian/rsnark/provers-gnark/prover/types"
)

type Groth16ProverCall struct{}

var (
	provers     []*prover.Groth16Prover
	proverMutex sync.Mutex
)

func init() {
	Groth16ProverImpl = Groth16ProverCall{}
}

func (p Groth16ProverCall) create(curve *uint64) uint64 {
	curveType := types.CurveType(*curve)

	prover := prover.NewGroth16Prover(curveType)
	proverMutex.Lock()
	defer proverMutex.Unlock()

	provers = append(provers, prover)

	return uint64(len(provers) - 1)
}

func (p Groth16ProverCall) compile(prover_id *uint64, circuitData *[]byte) int64 {
	circuitDef, err := circuit.ParseCircuitDefinition(*circuitData)
	if err != nil {
		log.Fatalf("failed to parse circuit definition: %v", err)
		return -20001
	}

	proverMutex.Lock()
	defer proverMutex.Unlock()

	prover := provers[*prover_id]
	compiled, err := prover.Compile(circuitDef)
	if err != nil {
		log.Fatalf("failed to compile circuit: %v", err)
		return -20002
	}

	objectMutex.Lock()
	defer objectMutex.Unlock()

	objects = append(objects, compiled)

	return int64(len(objects) - 1)
}

func (p Groth16ProverCall) setup(prover_id *uint64, compiled_circuit_id *int64) []byte {
	proverMutex.Lock()
	defer proverMutex.Unlock()

	objectMutex.Lock()
	defer objectMutex.Unlock()

	prover := provers[*prover_id]
	compiled, ok := objects[*compiled_circuit_id].(*types.CompiledCircuit)
	if !ok {
		log.Fatalf("failed to cast compiled circuit to types.CompiledCircuit")
		return int64ToBytes2(-20003, 0)
	}

	pk, vk, err := prover.Setup(compiled)

	if err != nil {
		log.Fatalf("failed to setup Groth16: %v", err)
		return int64ToBytes2(-20004, 0)
	}

	objects = append(objects, pk)
	res0 := int64(len(objects) - 1)

	objects = append(objects, vk)
	res1 := int64(len(objects) - 1)

	return int64ToBytes2(res0, res1)
}

func int64ToBytes2(i0 int64, i1 int64) []byte {
	bigEndian := make([]byte, 16)
	binary.BigEndian.PutUint64(bigEndian, uint64(i0))
	binary.BigEndian.PutUint64(bigEndian[8:], uint64(i1))

	return bigEndian
}

func int64ToBytes(i0 int64) []byte {
	bigEndian := make([]byte, 8)
	binary.BigEndian.PutUint64(bigEndian, uint64(i0))

	return bigEndian
}

func (p Groth16ProverCall) prove(prover_id *uint64, compiled_circuit_id *int64, pk_id *int64, witness_data *[]byte) []byte {
	proverMutex.Lock()
	defer proverMutex.Unlock()

	objectMutex.Lock()
	defer objectMutex.Unlock()

	prover := provers[*prover_id]
	pk, ok := objects[*pk_id].(*types.Groth16ProvingKey)
	if !ok {
		log.Fatalf("failed to cast pk to types.Groth16ProvingKey")
		return int64ToBytes(-20005)
	}

	compiled, ok := objects[*compiled_circuit_id].(*types.CompiledCircuit)
	if !ok {
		log.Fatalf("failed to cast compiled circuit to types.CompiledCircuit")
		return int64ToBytes(-20006)
	}

	var witness types.TemplateWitness
	err := witness.FromJSON(*witness_data)
	if err != nil {
		log.Fatalf("failed to deserialize witness: %v", err)
		return int64ToBytes(-20007)
	}

	proof, err := prover.Prove(compiled, pk, &witness)
	if err != nil {
		log.Fatalf("failed to prove: %v", err)
		return int64ToBytes(-20008)
	}

	code := int64ToBytes(0)
	return append(code, proof...)

}

func (p Groth16ProverCall) verify(prover_id *uint64, vk_id *int64, proof_data *[]byte, public_witness_data *[]byte) int64 {
	proverMutex.Lock()
	defer proverMutex.Unlock()

	objectMutex.Lock()
	defer objectMutex.Unlock()

	prover := provers[*prover_id]
	vk, ok := objects[*vk_id].(*types.Groth16VerifyingKey)
	if !ok {
		log.Fatalf("failed to cast vk to types.Groth16VerifyingKey")
		return -20009
	}

	var public_witness types.TemplatePublicWitness
	err := public_witness.FromJSON(*public_witness_data)
	if err != nil {
		log.Fatalf("failed to deserialize public witness: %v", err)
		return -20007
	}

	err = prover.Verify(*proof_data, vk, &public_witness)
	if err != nil {
		log.Fatalf("failed to verify: %v", err)
		return -20010
	}

	return int64(0)
}

File: provers/gnark/src/groth16.rs
use std::marker::PhantomData;

use rsnark_core::{
    CurveId, MetadataInfo, ProvingSystem, U256,
    types::{CircuitDefinition, PublicWitness, Witness},
};
use rsnark_provers_core::Backend;

use crate::{
    Error, Result, ffi,
    types::{CompiledCircuit, GoInnerRef, Groth16Proof, Groth16ProvingKey, Groth16VerifyingKey},
};

/// Groth16 backend implementation using the Gnark library.
///
/// This backend provides a Rust interface to Gnark's Groth16 implementation,
/// offering high-performance zero-knowledge proof generation and verification.
/// The backend manages Go-side resources through FFI and supports multiple
/// elliptic curves.
///
/// # Type Parameters
///
/// * `C` - The elliptic curve type that implements [`CurveId`]
///
/// # Resource Management
///
/// The backend maintains a reference to Go-side objects through `go_ref_id`.
/// These resources are managed by the Gnark library and cleaned up automatically
/// when no longer referenced.
pub struct Groth16Backend<C> {
    go_ref_id: u64,
    marker: PhantomData<C>,
}

impl<C> Drop for Groth16Backend<C> {
    fn drop(&mut self) {
        ffi::groth16::remove_prover(self.go_ref_id);
    }
}

impl<C> Clone for Groth16Backend<C>
where
    C: CurveId,
{
    fn clone(&self) -> Self {
        Self {
            go_ref_id: self.go_ref_id,
            marker: PhantomData,
        }
    }
}

impl<C> Groth16Backend<C>
where
    C: CurveId,
{
    fn _new() -> Self {
        let curve = C::curve_id();

        let prover = ffi::groth16::create(curve);

        Self {
            go_ref_id: prover,
            marker: PhantomData,
        }
    }

    fn _compile(&self, circuit: &CircuitDefinition) -> Result<CompiledCircuit<C>> {
        let circuit = serde_json::to_vec(circuit)?;

        let curve = C::curve_id();

        let res = ffi::groth16::compile(curve, circuit);

        if res >= 0 {
            Ok(CompiledCircuit::from_go_inner_ref(res))
        } else {
            Err(Error::from_go_error(res))
        }
    }

    fn _setup(
        &self,
        compiled_circuit: &CompiledCircuit<C>,
    ) -> Result<(Groth16ProvingKey<C>, Groth16VerifyingKey<C>)> {
        let res = ffi::groth16::setup(self.go_ref_id, compiled_circuit.go_inner_ref());

        let res0 = i64::from_be_bytes(res[0..8].try_into().unwrap());
        let res1 = i64::from_be_bytes(res[8..16].try_into().unwrap());

        if res0 >= 0 && res1 >= 0 {
            Ok((
                Groth16ProvingKey::from_go_inner_ref(res0),
                Groth16VerifyingKey::from_go_inner_ref(res1),
            ))
        } else {
            Err(Error::from_go_error(res0))
        }
    }

    fn _prove(
        &self,
        compiled_circuit: &CompiledCircuit<C>,
        pk: &Groth16ProvingKey<C>,
        witness: &Witness,
    ) -> Result<Groth16Proof<C>> {
        let witness_bytes = serde_json::to_vec(witness)?;

        let res = ffi::groth16::prove(
            self.go_ref_id,
            compiled_circuit.go_inner_ref(),
            pk.go_inner_ref(),
            witness_bytes,
        );

        if res < 0 {
            Err(Error::from_go_error(res))
        } else {
            Ok(Groth16Proof::from_go_inner_ref(res))
        }
    }

    fn _verify(
        &self,
        vk: &Groth16VerifyingKey<C>,
        proof: &Groth16Proof<C>,
        public_witness: &PublicWitness,
    ) -> Result<bool> {
        let public_witness_bytes = serde_json::to_vec(public_witness)?;

        let res = ffi::groth16::verify(
            self.go_ref_id,
            vk.go_inner_ref(),
            proof.go_inner_ref(),
            public_witness_bytes,
        );

        if res == 0 {
            Ok(true)
        } else if res == -20010 {
            Ok(false)
        } else {
            Err(Error::from_go_error(res))
        }
    }
}

/// Implementation of the [`Backend`] trait for Groth16 using Gnark.
///
/// This implementation bridges Rust's type system with Gnark's Go implementation,
/// providing compile-time curve selection and runtime proof operations.
impl<C> Backend for Groth16Backend<C>
where
    C: CurveId,
{
    type CircuitConstraint = CompiledCircuit<C>;
    type ProvingKey = Groth16ProvingKey<C>;
    type VerifyingKey = Groth16VerifyingKey<C>;
    type Proof = Groth16Proof<C>;

    type Error = Error;

    fn new() -> Self {
        Self::_new()
    }

    fn metadata(&self) -> MetadataInfo {
        MetadataInfo {
            field: C::field(),
            curve: C::curve_type(),
            proving_system: ProvingSystem::Groth16,
        }
    }

    fn compile(&self, circuit: &CircuitDefinition) -> Result<Self::CircuitConstraint> {
        self._compile(circuit)
    }

    fn setup(
        &self,
        compiled_circuit: &Self::CircuitConstraint,
    ) -> Result<(Self::ProvingKey, Self::VerifyingKey)> {
        self._setup(compiled_circuit)
    }

    fn prove(
        &self,
        compiled_circuit: &Self::CircuitConstraint,
        pk: &Self::ProvingKey,
        witness: &Witness,
    ) -> Result<Self::Proof> {
        let proof = self._prove(compiled_circuit, pk, witness)?;

        Ok(proof)
    }

    fn verify(
        &self,
        vk: &Self::VerifyingKey,
        proof: &Self::Proof,
        public_witness: &PublicWitness,
    ) -> Result<bool> {
        self._verify(vk, proof, public_witness)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rsnark_core::{API, Circuit, CircuitDefine, CircuitWitness, curve::BN254};
    use rsnark_provers_core::Prover;

    #[derive(rsnark_core::Circuit)]
    pub struct TestCircuit {
        a: u32,
        b: u32,
        pub c: u32,
    }

    impl Circuit for CircuitDefine<TestCircuit> {
        fn define(&self, api: &mut impl API) {
            let c = api.add(&self.a, &self.b);
            api.assert_is_equal(&c, &self.c);
        }
    }

    #[test]
    fn test_groth16_with_core_prover() {
        let prover: Prover<Groth16Backend<BN254>> = Prover::new();

        let circuit_prover = prover.compile_circuit::<TestCircuit>().unwrap();

        let (pk, vk) = circuit_prover.setup().unwrap();

        let circuit_witness = TestCircuit {
            a: 3,
            b: 4,
            c: 7, // 3 + 4 = 7
        };

        let proof = circuit_prover.prove(&pk, &circuit_witness).unwrap();

        let public_witness = circuit_witness.into_public_witness();
        circuit_prover.verify(&vk, &proof, public_witness).unwrap();
    }
}
